deduplication-document-spark/deduplication-document-spark-impl/src/main/scala/pl/edu/icm/coansys/document/deduplication/DeduplicateDocuments.scala

```diff
@@ -0,0 +1,420 @@
+ * 
+ */
+object DeduplicateDocuments {
+  val log = org.slf4j.LoggerFactory.getLogger(getClass().getName())
+
+  implicit def toJavaBiPredicate[A, B](predicate: (A, B) => Boolean) =
+    new BiPredicate[A, B] {
+      def test(a: A, b: B) = predicate(a, b)
+    }
+
+  def isValidDocument(doc: DocumentWrapper): Boolean = { //todo: fix based on if return value.
+    if (doc.hasDocumentMetadata()) {
+      val md = doc.getDocumentMetadata
+      if (md.hasBasicMetadata) {
+        val bmd = md.getBasicMetadata
+        (bmd.getTitleCount() > 0 || bmd.getAuthorCount > 0 || bmd.hasDoi || bmd.hasJournal)
+      } else {
+          false
+      }
+    } else {
+        false
+    }
+  }
+
+    
+  def calculateKeys(doc: DocumentMetadata, initialClusteringKeySize: Int, maximumClusteringKeySize: Int): Seq[String] = {
+    val keySizes = initialClusteringKeySize to maximumClusteringKeySize
+    var res = MultiLengthTitleKeyGenerator.generateKeys(doc)(keySizes)
+    if (res.head.isEmpty) {
+      res = Array.fill[String](keySizes.length)(doc.getKey)
+    }
+    res
+  }
+
+  /**
+   * Group items into large clusters, within which detailed analysis will be
+   * held. 
+   *
+   *  Items are grouped by keys generated from the normalised titles.
+   * If the cluster is too big, then longer keys are used, so smaller clusters are 
+   * generated. Treshold is maximumClusterSize.
+   * 
+   */
+  def prepareInitialClustering(inputDocs: RDD[(String, DocumentWrapper)], initialClusteringKeySize: Int,
+    maximumClusteringKeySize: Int, maximumClusterSize: Int): RDD[(String, Iterable[DocumentWrapper])] = {
+    log.info("Initializing cluster preparation.")
+    val keySizes = initialClusteringKeySize to maximumClusteringKeySize
+    log.info("Will use key sizes: " + keySizes.mkString(", "))
+    
+    val idClusterKeys = inputDocs.mapValues(doc => calculateKeys(
+            doc.getDocumentMetadata(), initialClusteringKeySize, maximumClusteringKeySize)) //we loose documents here, ony ids are preseved
+    val clusterDoc = idClusterKeys.flatMap(kv => kv._2.map(idcluster => (idcluster, kv._1))) // (clusterId => docId)
+    val clusterSizes = idClusterKeys.flatMap(x => (x._2.map(y => (y, 1)))).reduceByKey(_ + _) //(clusterId => clusterSize)
+    
+    //build rdd (docId, (clusterId, clusterSize) )
+    val docClustersWithSizes = clusterDoc.join(clusterSizes).map(p => (p._2._1, (p._1, p._2._2)))
+    //build rdd - (docId, clusterId)
+    val selectedClusters = docClustersWithSizes.reduceByKey((x, y) => {
+      if (x._2 <= maximumClusterSize) {
+        if (y._2 <= maximumClusterSize) {
+          if (x._1.length <= y._1.length) { x } else { y }
+        } else {
+          x
+        }
+      } else {
+        if (y._2 <= maximumClusterSize) {
+          y
+        } else {
+          if (x._1.length > y._1.length) { x } else { y }
+        }
+      }
+    }).mapValues(_._1)
+    inputDocs.join(selectedClusters).map(p => (p._2._2, p._2._1)).groupByKey
```

Comment:
Last expression could be simplified to:
`selectedClusters.join(inputDocs).values.groupByKey`